botorch/acquisition/multi_objective/predictive_entropy_search.py

```diff
@@ -0,0 +1,1629 @@
+            cov_c_new = torch.cholesky_inverse(cholesky_nat_cov_c_new)
+            # cov_c_new = torch.inverse(nat_cov_c_new + self.jitter * identity_c)
+            check_no_nans(cov_c_new)
+
+        ###########################################################################
+        # COMPUTE THE LOG DETERMINANTS
+        ###########################################################################
+        # compute the initial log determinant term
+        log_det_pred_f_cov_noise = _compute_log_determinant(pred_f_cov_noise, q)
+        # compute the post log determinant term
+        log_det_cov_f = _compute_log_determinant(cov_f_new + observation_noise, q)
+        if K > 0:
+            # compute the initial log determinant term
+            log_det_pred_c_cov_noise = _compute_log_determinant(pred_c_cov_noise, q)
+            # compute the post log determinant term
+            log_det_cov_c = _compute_log_determinant(
+                cov_c_new + observation_noise_c, q
+            )
+        ###########################################################################
+        # COMPUTE THE ACQUISITION FUNCTION
+        ###########################################################################
+        q_pes_f = log_det_pred_f_cov_noise - log_det_cov_f
+        check_no_nans(q_pes_f)
+        if K > 0:
+            q_pes_c = log_det_pred_c_cov_noise - log_det_cov_c
+            check_no_nans(q_pes_c)
+        else:
+            q_pes_c = 0
+
+        return .5 * (q_pes_f + q_pes_c)
+
+
+def log_cdf_robust(x: Tensor) -> Tensor:
+    r""" Computes the logarithm of the normal cumulative density robustly. This uses
+    the approximation log(1-z) ~ -z when z is small:
+    if x > 5:
+        log(cdf(x)) = log(1-cdf(-x)) approx -cdf(-x)
+    else:
+        log(cdf(x)).
+
+    Args:
+        x: a `x_shape`-dim Tensor.
+
+    Returns
+        A `x_shape`-dim Tensor.
+    """
+    normal = Normal(torch.zeros_like(x), torch.ones_like(x))
+    cdf_x = normal.cdf(x)
+    neg_cdf_neg_x = - normal.cdf(-x)
+    log_cdf_x = torch.where(x < 5, torch.log(cdf_x), neg_cdf_neg_x)
+
+    return log_cdf_x.clamp(NEG_INF, -CLAMP_LB)
+
+
+def _initialize_predictive_matrices(
+        X: Tensor,
+        model: Model,
+        observation_noise: Optional[bool] = True,
+        jitter: Optional[float] = 1e-4,
+        natural: Optional[bool] = True,
+) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
+    r""" Initializes the natural predictive mean and covariance matrix. For a
+    multivariate normal distribution with mean mu and covariance Sigma, the natural
+    mean is Sigma^{-1} mu and the natural covariance is Sigma^{-1}.
+
+    Args:
+        X: A `batch_shape x R x d`-dim Tensor.
+        model: The fitted model.
+        observation_noise: If true the posterior is computed with observation noise.
+        jitter: The jitter added to the covariance matrix.
+        natural: If true we compute the natural statistics as well.
+    Return:
```

Comment:
Probably need to follow this format for the autodocs: https://github.com/pytorch/botorch/blob/970664c2d8516dca7c583858a10794f78c84020f/botorch/optim/optimize.py#L110-L118